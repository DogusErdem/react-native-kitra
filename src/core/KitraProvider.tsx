import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
