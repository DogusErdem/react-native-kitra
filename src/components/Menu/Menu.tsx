import { useCallback, useEffect, useRef, useState } from 'react';
import { Dimensions, StyleSheet, Text, View } from 'react-native';
import Animated, { FadeIn, FadeOut } from 'react-native-reanimated';
import { TouchableOpacity } from 'react-native-gesture-handler';
import type { FCCWD, MenuProps } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';
import Divider from '../Divider/Divider';

const HEIGHT = Dimensions.get('window').height;

const styles = StyleSheet.create({
  container: {
    zIndex: 1,
  },
  openButton: {
    alignItems: 'flex-end',
    marginRight: 10,
  },
  menuContainer: {
    width: 200,
    borderRadius: 5,
    position: 'absolute',
  },
  menuButton: {
    margin: 10,
    flexDirection: 'row',
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    alignItems: 'center',
  },
});

const Menu : FCCWD<MenuProps> = (
  { typography,
    theme,
    items,
    dividerColor,
    menuStyle,
    labelStyle,
    containerStyle,
    rowStyle,
    closeOnPress,
    button = () => <Text style={{ color: theme?.primary, fontSize: 20, bottom: 5 }}>...</Text> },
) => {
  const [open, setOpen] = useState(false);
  const [size, setSize] = useState({ x: 0, y: 0, width: 0, height: 0 });
  const [menuHeight, setMenuHeight] = useState({ height: 0 });
  const menu = useRef<View>(null);

  useEffect(() => {
    menu.current?.measure((x, y, width, height, pageX, pageY) => {
      setSize({ x: pageX, y: pageY, width, height });
    });
  }, [open]);

  const closeMenu = useCallback(() => {
    if (closeOnPress) {
      setOpen(false);
    }
  }, [closeOnPress]);
  return (
    <View style={[styles.container, { width: menuStyle ? menuStyle?.width : 200 }, containerStyle]} ref={menu}>
      <TouchableOpacity testID="open_button" onPress={() => setOpen(!open)} style={styles.openButton}>
        {button(open)}
      </TouchableOpacity>
      {open ? (
        <Animated.View
          testID="menu_container"
          style={[
            styles.menuContainer,
            { backgroundColor: menuStyle?.backgroundColor, right: 0 },
            HEIGHT - (size.y + menuHeight.height + size.height) >= 0 ? { top: size.height + 10 } : { bottom: 30 }, menuStyle,
          ]}
          entering={FadeIn.duration(300)}
          exiting={FadeOut.duration(300)}
          onLayout={e => setMenuHeight({ height: e.nativeEvent.layout.height })}
        >
          {items.map((item, index) => (
            <View key={index} style={rowStyle}>
              <TouchableOpacity testID={`item_button_${index}`} activeOpacity={closeOnPress ? 0.4 : 1} onPress={() => { item.onPress && item.onPress(); closeMenu(); }} style={[styles.menuButton, { backgroundColor: menuStyle?.backgroundColor }]}>
                {item.left && item.left}
                <View style={styles.buttonContainer}>
                  <Text style={[
                    { fontSize: typography?.body.smedium.fontSize,
                      fontWeight: '500',
                      lineHeight: typography?.body.smedium.lineHeight,
                      paddingHorizontal: item.left ? 5 : 0,
                      color: theme?.lightBlack }, labelStyle]}
                  >
                    {item.label}
                  </Text>
                  {item.right && item.right}
                </View>
              </TouchableOpacity>
              {items.length - 1 !== index && dividerColor ?
                <Divider theme={theme} typography={typography} color={dividerColor} width={Number(menuStyle?.width) - 20 || 180} />
                : null}
            </View>
          ))}
        </Animated.View>
      )
        :
        null
    }
    </View>

  );
};

export default applyDefaults(Menu);
