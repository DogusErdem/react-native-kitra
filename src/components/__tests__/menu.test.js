/* eslint-disable no-undef */

import { fireEvent, render, waitFor, screen, act } from '@testing-library/react-native';
import { Text } from 'react-native';

import { KitraProvider } from '../../core/KitraProvider';
import Menu from '../Menu/Menu';

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('Menu', () => {
  describe('renders without crashing', () => {
    const { toJSON } = renderWithContext(<Menu items={[]} />);
    expect(toJSON()).toMatchSnapshot();
  });
  describe('open and close menu', () => {
    const onPress = jest.fn();
    const onPress1 = jest.fn();
    const onPress2 = jest.fn();
    it('toggles open on button press', async () => {
      renderWithContext(<Menu
        closeOnPress
        items={[
          { label: 'test', left: <Text>sada</Text>, onPress },
          { label: 'test', right: <Text>test</Text>, left: <Text>test</Text>, dividerColor: 'red', onPress: onPress1 },
          { label: 'test', right: <Text>test</Text>, left: <Text>test</Text>, onPress: onPress2 },
        ]}
      />);
      const openButton = screen.getByTestId('open_button');
      fireEvent.press(openButton);

      const menuContainer = screen.getByTestId('menu_container');
      expect(menuContainer).toBeVisible();

      const closeButton = screen.getByTestId('item_button_0');
      fireEvent.press(closeButton);

      expect(onPress).toHaveBeenCalled();
    });
    it('menu close with open button', async () => {
      renderWithContext(<Menu
        items={[
          { label: 'test', left: <Text>sada</Text>, onPress },
          { label: 'test', right: <Text>test</Text>, left: <Text>test</Text>, dividerColor: 'red', onPress: onPress1 },
          { label: 'test', right: <Text>test</Text>, left: <Text>test</Text>, onPress: onPress2 },
        ]}
      />);
      const openButton = screen.getByTestId('open_button');
      fireEvent.press(openButton);
      const menuContainer = screen.getByTestId('menu_container');
      act(() => {
        fireEvent(menuContainer, 'layout', {
          nativeEvent: { layout: { } },
        });
      });
      expect(menuContainer).toBeVisible();

      fireEvent.press(openButton);
    });
  });
});
