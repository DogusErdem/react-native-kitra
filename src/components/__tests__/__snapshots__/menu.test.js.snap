// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "zIndex": 1,
        },
        {
          "width": 200,
        },
        undefined,
      ]
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      delayLongPress={600}
      enabled={true}
      exclusive={true}
      handlerTag={1}
      handlerType="NativeViewGestureHandler"
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      testID="open_button"
      touchSoundDisabled={false}
    >
      <View
        accessible={true}
        collapsable={false}
        style={
          {
            "alignItems": "flex-end",
            "marginRight": 10,
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            {
              "bottom": 5,
              "color": "#8973CD",
              "fontSize": 20,
            }
          }
        >
          ...
        </Text>
      </View>
    </RNGestureHandlerButton>
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
